crates/winch/src/codegen.rs

```diff
@@ -0,0 +1,221 @@
+    }
+}
+
+/// The code generation abstraction
+pub(crate) struct CodeGen<'c, 'a: 'c, M>
+where
+    M: MacroAssembler,
+{
+    /// A reference to the function body
+    function: &'c mut FunctionBody<'a>,
+
+    /// The word size information, extracted from the current ABI
+    word_size: u32,
+
+    /// The ABI-specific representation of the function signature, excluding results
+    sig: ABISig,
+
+    /// The code generation context
+    pub context: CodeGenContext<'c, M>,
+
+    /// The register allocator
+    pub regalloc: RegAlloc,
+
+    /// Function body validator
+    pub validator: &'a mut FuncValidator<ValidatorResources>,
+}
+
+impl<'c, 'a: 'c, M> CodeGen<'a, 'c, M>
+where
+    M: MacroAssembler,
+{
+    pub fn new(
+        context: CodeGenContext<'c, M>,
+        word_size: u32,
+        sig: ABISig,
+        function: &'c mut FunctionBody<'a>,
+        validator: &'c mut FuncValidator<ValidatorResources>,
+        regalloc: RegAlloc,
+    ) -> Self {
+        Self {
+            function,
+            word_size,
+            sig,
+            context,
+            regalloc,
+            validator,
+        }
+    }
+
+    /// Emit the function body to machine code
+    pub fn emit(&mut self) -> Result<Vec<String>> {
+        self.emit_start()
+            .and(self.emit_body())
+            .and(self.emit_end())?;
+        let buf = self.context.masm.finalize();
+        let code = Vec::from(buf);
+        Ok(code)
+    }
+
+    // TODO stack checks
+    fn emit_start(&mut self) -> Result<()> {
+        self.context.masm.prologue();
+        self.context
+            .masm
+            .reserve_stack(self.context.frame.locals_size);
+        Ok(())
+    }
+
+    fn emit_body(&mut self) -> Result<()> {
+        self.spill_register_arguments();
+        self.zero_local_slots();
+
+        let mut ops_reader = self.function.get_operators_reader()?;
+        while !ops_reader.eof() {
+            ops_reader.visit_with_offset(self)??;
+        }
+        ops_reader.ensure_end().map_err(|e| e.into())
+    }
+
+    // Emit the usual function end instruction sequence
+    pub fn emit_end(&mut self) -> Result<()> {
+        self.handle_abi_result();
+        self.context.masm.epilogue(self.context.frame.locals_size);
+        Ok(())
+    }
+
+    fn zero_local_slots(&mut self) {
+        let range = &self.context.frame.defined_locals_range;
+        if range.0.start() == range.0.end() {
+            return;
+        }
+
+        // Divide the locals range into word-size slots; first ensure that the range limits
+        // are word size aligned; since there's no guarantee about their alignment. The aligned "upper"
+        // limit should always be less than or equal to the size of the local area, which gets
+        // validated when getting the address of a local
+
+        let word_size = self.word_size;
+        // If the locals range start is not aligned to the word size, zero the last four bytes
+        let range_start = range
```

Comment:
Can we factor out this logic into a `memset` helper on the `MacroAssembler`, that either unrolls a short sequence of stores or calls a helper?

Reply:
I think it's reasonable to do this since this is common for all ISAs and will be cleaner if we put this in the masm. A few clarification questions: 

1. With this logic you mean the entire piece to zero out a locals range?
2. By helper, is your idea to have this in the masm module, but not necessarily part of the trait?

Reply:
Right, the entire bit to zero a range of memory; and I think it'd be pretty reasonable to have it as a trait method on `MacroAssembler` that has a default body (which does some basic lowering to either unrolled stores or a loop that will work on any architecture). That also leaves room for more efficient implementations per machine later (e.g. `rep movsb` on x86 machines with [ERMS](https://stackoverflow.com/questions/43343231/enhanced-rep-movsb-for-memcpy) where that's fast, or something using SIMD, or ...).

Reply:
Makes sense to me; I've factored this out into a method in the trait with a default implementation. I've also made the code a bit simpler.